src/clusterfuzz/_internal/cron/schedule_fuzz.py

```diff
@@ -0,0 +1,227 @@
+
+
+def get_available_cpus(project: str, region: str) -> int:
+  """Gets the number of available CPUs in the current GCE region."""
+  quotas = _get_quotas(project, region)
+
+  # If preemptible quota is not defined, we need to use CPU quota instead.
+  cpu_quota = None
+  preemptible_quota = None
+
+  # Get preemptible_quota and cpu_quota from the list of quotas.
+  for quota in quotas:
+    if preemptible_quota and cpu_quota:
+      break
+    if quota['metric'] == 'CPUS':
+      cpu_quota = quota
+      continue
+    if quota['metric'] == 'PREEMPTIBLE_CPUS':
+      preemptible_quota = quota
+      continue
+  assert preemptible_quota or cpu_quota
+
+  if not preemptible_quota['limit']:
+    quota = cpu_quota
+  else:
+    quota = preemptible_quota
+  assert quota['limit'], quota
+
+  return quota['limit'] - quota['usage']
+
+
+class BaseFuzzTaskScheduler:
+  """Base fuzz task scheduler for any deployment of ClusterFuzz."""
+
+  def __init__(self, num_cpus):
+    self.num_cpus = num_cpus
+
+  def get_fuzz_tasks(self):
+    raise NotImplementedError('Child class must implement.')
+
+  def _get_cpus_per_fuzz_job(self, job_name):
+    del job_name
+    # TODO(metzman): Actually implement this.
+    return 2
+
+
+class FuzzerJob:
+  """Data class that holds more info about FuzzerJobs than the ndb.Models do.
+  Something like this would probably not be needed if we were using SQL and
+  could use joins."""
+
+  def __init__(self, job, project, queue, fuzzer=None, weight=None):
+    self.job = job
+    self.project = project
+    self.queue = queue
+    self.fuzzer = fuzzer
+    self.weight = weight
+
+  def copy(self):
+    return FuzzerJob(
+        job=self.job,
+        project=self.project,
+        queue=self.queue,
+        fuzzer=self.fuzzer,
+        weight=self.weight)
+
+
+class OssfuzzFuzzTaskScheduler(BaseFuzzTaskScheduler):
+  """Fuzz task scheduler for OSS-Fuzz."""
+
+  def get_fuzz_tasks(self) -> Dict[str, tasks.Task]:
+    # TODO(metzman): Handle high end.
+    # A job's weight is determined by its own weight and the weight of the
+    # project is a part of. First get project weights.
+    logs.info('Getting projects.')
+    projects = list(
+        ndb_utils.get_all_from_query(data_types.OssFuzzProject.query()))
+
+    total_cpu_weight = sum(project.cpu_weight for project in projects)
+    project_weights = {}
+    for project in projects:
+      project_weight = project.cpu_weight / total_cpu_weight
+      project_weights[project.name] = project_weight
+
+    # Then get FuzzerJob weights.
+    logs.info('Getting jobs.')
+    jobs = {}
+    for job in ndb_utils.get_all_from_query(data_types.Job.query()):
+      jobs[job.name] = FuzzerJob(
+          job=job.name,
+          project=job.project,
+          queue=tasks.queue_for_platform(job.platform))
+
+    fuzzer_job_weight_by_project = collections.defaultdict(int)
+    fuzzer_jobs = {}
+    fuzzer_job_query = ndb_utils.get_all_from_query(
+        data_types.FuzzerJob.query())
```

Comment:
Is this performant enough? Do we need to use `FuzzerJobs` (plural) here? (Which IIRC we created to address performance issues).

Reply:
It is very performant. I'm not sure what the original perf issue was.
Locally (where I have a bunch of cores, thus improving the speed of the actual bottleneck, putting the tasks on the queue) we can schedule ~125 tasks in ~15 seconds.